libdrgn/dwarf_index.c

```diff
@@ -2074,6 +2099,8 @@ static struct drgn_error *index_die(struct drgn_dwarf_index_namespace *ns,
 	struct drgn_dwarf_index_die *die;
 
 	hp = drgn_dwarf_index_die_map_hash(&entry.key);
+	if ((err = drgn_dwarf_index_namespace_shards_init(ns)))
+		return err;
```

Comment:
I think it'd be simpler to set up the shards before we start indexing (in `drgn_dwarf_index_update()` and `index_namespace()`) rather than checking on every call to `index_die()`. I guess doing it that way would allocate unnecessary memory if the namespace is truly empty, but I don't think we need to be concerned with that for a few reasons:

1. The memory still won't be allocated until the user attempts to do a lookup in the namespace (which they probably won't do seeing as there is nothing in that namespace).
2. Neither Clang nor GCC seems to generate debug info for empty namespace definitions anyways.
3. If you really wanted to, you could have an extra check after indexing that frees the shards if they're all empty.

Reply:
Yeah, the main concern I'm trying to fix is namespaces which haven't been requested yet - setting up shards in those functions seems fine to me!